modules/imgproc/src/warp_kernels.simd.hpp

```diff
@@ -2130,6 +1592,1414 @@ void warpAffineLinearApproxInvoker_8UC4(const uint8_t *src_data, size_t src_step
+                            srcrows <= 1 ? BORDER_REPLICATE : border_type;
+
+#if (CV_SIMD || CV_SIMD_SCALABLE)
+        constexpr int max_vlanes_32{VTraits<v_float32>::max_nlanes};
+        constexpr int max_uf{max_vlanes_32*2};
+        int vlanes_32 = VTraits<v_float32>::vlanes();
+        // unrolling_factor = lane_size / 16 = vlanes_32 * 32 / 16 = vlanes_32 * 2
+        int uf = vlanes_32 * 2;
+
+        std::array<float, max_vlanes_32> start_indices;
+        std::iota(start_indices.data(), start_indices.data() + max_vlanes_32, 0.f);
+
+        v_uint32 inner_srows = vx_setall_u32((unsigned)srcrows - 2),
+                 inner_scols = vx_setall_u32((unsigned)srccols - 1),
+                 outer_srows = vx_setall_u32((unsigned)srcrows + 1),
+                 outer_scols = vx_setall_u32((unsigned)srccols + 1);
+        v_float32 delta = vx_setall_f32(static_cast<float>(uf));
+        v_int32 one = vx_setall_s32(1), four = vx_setall_s32(4);
+        v_int32 v_srcstep = vx_setall_s32(int(srcstep));
+        int32_t addr[max_uf],
+                src_ix[max_uf],
+                src_iy[max_uf];
+        uint16_t pixbuf[max_uf*4*4];
+
+        uint16_t bvalbuf[max_uf*4];
+        for (int i = 0; i < uf; i++) {
+            bvalbuf[i*4] = bval[0];
+            bvalbuf[i*4+1] = bval[1];
+            bvalbuf[i*4+2] = bval[2];
+            bvalbuf[i*4+3] = bval[3];
+        }
+        v_uint16 bval_v0 = vx_load(&bvalbuf[0]);
+        v_uint16 bval_v1 = vx_load(&bvalbuf[uf]);
+        v_uint16 bval_v2 = vx_load(&bvalbuf[uf*2]);
+        v_uint16 bval_v3 = vx_load(&bvalbuf[uf*3]);
+#endif
+
+        for (int y = r.start; y < r.end; y++) {
+            uint16_t* dstptr = dst + y*dststep;
+            int x = 0;
+
+#if (CV_SIMD || CV_SIMD_SCALABLE)
+            CV_WARPPERSPECTIVE_LINEAR_VECTOR_COMPUTE_MAPPED_COORD1();
+
+            for (; x < dstcols - uf; x += uf) {
+                // [TODO] apply halide trick
+
+                CV_WARPPERSPECTIVE_LINEAR_VECTOR_COMPUTE_MAPPED_COORD2(C4);
+
+                if (v_reduce_min(inner_mask) != 0) { // all loaded pixels are completely inside the image
+                    CV_WARP_LINEAR_VECTOR_SHUFFLE_ALLWITHIN(C4, 16U);
+                } else {
+                    CV_WARP_LINEAR_VECTOR_SHUFFLE_NOTALLWITHIN(C4, 16U);
+                }
+
+                CV_WARP_LINEAR_VECTOR_INTER_LOAD_U16(C4);
+
+                CV_WARP_LINEAR_VECTOR_INTER_LOAD_U16F32(C4);
+
+                CV_WARP_LINEAR_VECTOR_INTER_CALC_F32(C4);
+
+                CV_WARP_LINEAR_VECTOR_INTER_STORE_F32U16(C4);
+            }
+#endif // (CV_SIMD || CV_SIMD_SCALABLE)
+
+            for (; x < dstcols; x++) {
+                float w = x*M[6] + y*M[7] + M[8];
+                float sx = (x*M[0] + y*M[1] + M[2]) / w;
+                float sy = (x*M[3] + y*M[4] + M[5]) / w;
+                int ix = cvFloor(sx), iy = cvFloor(sy);
+                sx -= ix; sy -= iy;
+                int p00r, p00g, p00b, p00a, p01r, p01g, p01b, p01a;
```

Comment:
why declaration of variables is put outside of the macros?

Reply:
Ok, put inside the shuffle macro.